[JoanneHing] iP

JoanneHing opened the pull request:
# Toothless ⚫ 
> “What you’re searching for isn’t out there, Hiccup. It’s in here.” —Astrid, How To Train Your Dragon 2 ([source](https://www.sideshow.com/blog/how-to-train-your-dragon-quotes))

Toothless helps you to note down your quest so you can focus more on your adventures! It is
- Text-based 
- Easy to use 
- ~~FAST~~ **__SUPER FAST__** interaction
- It is also **FUN** to interact

How easy is it to use? You just need to
1. Type your command ⌨️ 
2. Toothless will note down everything for you! 💯 

The features available are
- [X] Add task (To-do, Deadline, Event)
- [x] Mark a task and done or undone
- [x] Remove a task
- [x] Delete a task
- [x] Find specific task
- [x] List down the tasks added

Some example code
```java
public void addTask(Task task, Ui ui, TaskList taskList) {
        list.add(task);
        ui.addTaskMessage(task, taskList.getList().size());
    }
```

Some commands available are
```java
todo <taskName> // Add a to-do task
deadline <taskName> /by <Date and Time> // Add a task with deadline
event <taskName> /from <Date and Time> /to <Date and Time> // Add an event with the date and time
mark <index> // Mark task as done
unmark <index> // Mark task as undone
delete <index> // Delete a task
find <keyword> // Find task with the keyword
list // Display the list of tasks
```
Example usage is 'todo study' that add study as a To-Do task in the list.

The best thing is, it is **FREE** for use!
Download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/index.html)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

JoanneHing committed:
feat: Implemented initial skeletal version with greetings and exit message

JoanneHing committed:
Implement initial skeletal version with greetings and exit message.

JoanneHing committed:
Merge branch 'master' of https://github.com/JoanneHing/ip

JoanneHing committed:
feat: Implement echoing and exit message feature

JoanneHing committed:
feat: Add the ability to store user-entered text and display it on request

JoanneHing committed:
feat: Add the ability to mark tasks as done and undone

JoanneHing committed:
feat: Add support for tracking three types of tasks: deadline, todos and events

JoanneHing committed:
fix: Fix display bug for the task type Events

JoanneHing committed:
test: Automated Text UI Testing

JoanneHing committed:
feat: Handle errors for invalid commands

JoanneHing committed:
feat: Add delete task command

JoanneHing committed:
refractor: Add enum CommandType.java

JoanneHing committed:
doc: fix the header comments

JoanneHing committed:
feat: Add saving tasks and loading task

JoanneHing committed:
feat: Implement date and time detection in task type Deadline and Event

LemonDrew commented on src/main/java/CommandType.java:
```diff
@@ -0,0 +1,39 @@
+     */
+    public static CommandType getCommandType(String command) {
+        switch (command.toLowerCase()) {
+            case "todo":
```
Perhaps there should be no indentation after the switch statement?

LemonDrew commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+     * @throws ToothlessExceptions if the date and time format is invalid
+     */
+    public Deadline(String description, String deadline) throws ToothlessExceptions{
+        super(description);
```
I like how you make good use of inheritance.

LemonDrew commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+        try {
+            this.deadline = LocalDateTime.parse(deadline.trim().replace("-","/"), INPUT_FORMATTER);
+        } catch (DateTimeParseException e) {
+            throw new ToothlessExceptions("Please enter a valid date and time\n" +
```
Is there a better name other than ToothlessExceptions?

LemonDrew commented on src/main/java/MissingIndex.java:
```diff
@@ -0,0 +1,18 @@
+ * MissingIndex class is a subclass of ToothlessExceptions.
+ * It is used to handle the case when the user does not provide an index for the task.
+ */
+public class MissingIndex extends ToothlessExceptions{
```
I like the way you categorize the exceptions.

LemonDrew commented on src/main/java/ToDos.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Represents a ToDos task.
+ */
+public class ToDos extends Task {
```
Perhaps Todo would be much better since there is only a single Todo task rather than multiple ones?

LemonDrew commented on src/main/java/Toothless.java:
```diff
@@ -0,0 +1,339 @@
+        String description = splitInput.length < 2? "" : splitInput[1];
+
+        switch(command) {
+            case TODO :
```
Missing handling of cases?

LemonDrew commented on src/main/java/Toothless.java:
```diff
@@ -0,0 +1,339 @@
+        System.out.println("Toothless:\n" +
+                "Greetings, Dragon Rider!\n\n" +
+                "I'm Toothless, your friendly dragon companion.\n" +
+                "What adventure shall we embark on today?\n\n" + DIVIDER);
```
I now know where ToothlessException comes from! But for readability, it would be much better to rename the exception.

LemonDrew commented on src/main/java/Toothless.java:
```diff
@@ -0,0 +1,339 @@
+        Toothless toothless = new Toothless();
+        try {
+            toothless.beginChat();
+        } catch (ToothlessExceptions e) {
```
Since you have already defined the subclasses of Toothless Exceptions, maybe it would be much better to handle the subclasses before the parent class instead?

LemonDrew reviewed (commented):
Overall, the code that you have made is enjoyable. One suggestion is to perhaps create separate classes for the Parser function and Storage function for better handling and readability.

JoanneHing committed:
refractor: Refractor closely related code as classes

JoanneHing committed:
refractor: Divide classes into packages

JoanneHing committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

JoanneHing committed:
feat: Add gradle to the project

JoanneHing committed:
test: Add JUnit tests to test the behavior of the code

JoanneHing committed:
doc: Add Javadoc comments

JoanneHing committed:
fix: Fix parse date and time error

JoanneHing committed:
doc: Update some JavaDoc to the overload constructor of deadline and event

Dominic-Khoo commented on src/main/java/toothless/ui/Ui.java:
```diff
@@ -0,0 +1,96 @@
+     */
+    public void welcome() {
+        System.out.println("Welcome to the dragon's den!\n" + TOOTHLESS_LOGO);
+        System.out.println("Toothless:\n" +
+                "Greetings, Dragon Rider!\n\n" +
+                "I'm Toothless, your friendly dragon companion.\n" +
+                "What adventure shall we embark on today?\n\n" + DIVIDER);
```
Perhaps the line break should be done before the operator to improve readability?

Dominic-Khoo commented on src/main/java/toothless/ui/Ui.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    /**
+     * Prints the message when a task is added.
+     * @param task The task that is added.
+     * @param size The size of the task list.
```
Should there be an empty line between the parameter and description section?

Dominic-Khoo commented on src/main/java/toothless/command/Parser.java:
```diff
@@ -0,0 +1,87 @@
+        String[] splitCommand = trimmedCommand.split(" ", 2);
+        String description = splitCommand.length < 2? "" : splitCommand[1];
+
+        switch (splitCommand[0]) {
+        case "todo":
```
I like that the coding standard for switch and case was adhered to here!

Dominic-Khoo commented on src/main/java/toothless/storage/Storage.java:
```diff
@@ -0,0 +1,102 @@
+import toothless.task.Task;
+import toothless.task.ToDo;
+
+import java.io.*;
```
Perhaps the individual classes should be imported explicitly instead of the whole library?

Dominic-Khoo commented on src/main/java/toothless/command/ToDoCommand.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    public void executeCommand(TaskList taskList, Ui ui, Storage storage) throws ToothlessExceptions {
+        if(description.isEmpty()) {
```
Perhaps a white space could be added after "if"?

Dominic-Khoo reviewed (commented):
LGTM! Just a few coding standard nits to fix 😀

JoanneHing committed:
style: Fix code formatting and style

JoanneHing committed:
feat: Add find task feature

JoanneHing committed:
Merge branch 'Level-9'

* Level-9:
  feat: Add find task feature

JoanneHing committed:
fix: Add JavaDoc

JoanneHing committed:
fix: Fix message print for find task

JoanneHing committed:
style: fix code style using style check

JoanneHing committed:
doc: Update JavaDoc

JoanneHing committed:
Fix: Fix merge conflicts

JoanneHing committed:
doc: Add JavaDoc

JoanneHing committed:
Merge commit '4a271c8df53ebb7605bf614be5d3fc321af6f2cd'

* commit '4a271c8df53ebb7605bf614be5d3fc321af6f2cd':
  fix: Add JavaDoc

JoanneHing committed:
Merge tag 'A-CodingStandard' into A-JavaDoc

* tag 'A-CodingStandard':
  style: Fix code formatting and style

# Conflicts:
#	src/main/java/toothless/exceptions/ToothlessExceptions.java
#	src/main/java/toothless/task/Deadline.java
#	src/main/java/toothless/task/Event.java

JoanneHing committed:
Merge commit 'b9b526cf45c2eeb43cb8d419c9dbfdae6ac371f2'

* commit 'b9b526cf45c2eeb43cb8d419c9dbfdae6ac371f2':
  doc: Update JavaDoc

JoanneHing committed:
Merge commit '8a7ef86b39c7323b2cbcee0cb88255c43df97297'

* commit '8a7ef86b39c7323b2cbcee0cb88255c43df97297':
  doc: Add JavaDoc

# Conflicts:
#	src/main/java/toothless/exceptions/ToothlessExceptions.java

JoanneHing committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:

JoanneHing committed:
feat: Add GUI to the chatbot

JoanneHing committed:
Merge branch 'branch-A-GUI'

* branch-A-GUI:
  feat: Add GUI to the chatbot

JoanneHing committed:
Merge branch 'branch-A-GUI' into branch-Level-10

* branch-A-GUI:
  feat: Add GUI to the chatbot
  doc: Add JavaDoc
  doc: Update JavaDoc
  fix: Add JavaDoc
  doc: Update some JavaDoc to the overload constructor of deadline and event

JoanneHing committed:
Merge branch 'branch-Level-10'

* branch-Level-10:

JoanneHing committed:
feat: Add varargs at find method in TaskList.java

JoanneHing committed:
feat: add assertions to validate method parameters and internal states

Assertions are added to ensure that method parameters and internal
states are validated during development. This helps catch potential
issues early.

This change ensures that the code is more robust by catching null
values and invalid states early in the execution.

JoanneHing committed:
fix: improve code quality by refactoring

Refactored the code to enhance readability and maintainability.

These improvements make the codebase more robust and easier to
maintain.

JoanneHing committed:
feat: refactor findTask method to use Java Streams

Refactored the findTask method in TaskList.java to use Java Streams
for improved readability and performance.

These improvements make the code more concise and easier to
understand.

JoanneHing committed:
Merge pull request #1 from JoanneHing/branch-A-Assertions

Add assertions to validate task operations

JoanneHing committed:
Merge pull request #2 from JoanneHing/branch-A-CodeQuality

Improve code quality by refactoring

JoanneHing committed:
Merge branch 'master' into branch-A-Streams

JoanneHing committed:
Merge pull request #3 from JoanneHing/branch-A-Streams

Refactor findTask method to use Java Streams

JoanneHing committed:
Merge branch 'branch-A-GUI' of /Users/joannehing/Documents/CS2103T/Individual Project with conflicts.

JoanneHing committed:
feat: Improve the GUI of the chatbot

The paddings, fonts, colors and alignments are enhanced to make the GUI
more pleasant to look at.

JoanneHing committed:
Merge branch 'branch-A-BetterGUI' into branch-A-GUI

* branch-A-BetterGUI:
  feat: Improve the GUI of the chatbot

JoanneHing committed:
feat: Add help command to guide user how to use the command

JoanneHing committed:
Merge branch 'branch-A-GUI' into branch-A-BetterGUI

* branch-A-GUI:

JoanneHing committed:
feat: Improve anchor pane edges of the elements

JoanneHing committed:
fix: Fix merge conflict

JoanneHing committed:
Merge pull request #6 from JoanneHing/branch-A-BetterGUI

Enhance the GUI

JoanneHing committed:
fix: Fix merge conflict

JoanneHing committed:
Merge commit '87ef36032596bad64d43601eb72ded6a4c37a986'

* commit '87ef36032596bad64d43601eb72ded6a4c37a986':
  feat: Improve anchor pane edges of the elements
  feat: Improve the GUI of the chatbot
  Merge branch 'branch-A-GUI' of /Users/joannehing/Documents/CS2103T/Individual Project with conflicts.

JoanneHing committed:
Merge commit '87ef36032596bad64d43601eb72ded6a4c37a986' into A-BCD-Extension

* commit '87ef36032596bad64d43601eb72ded6a4c37a986':
  feat: Improve anchor pane edges of the elements

JoanneHing committed:
Merge branch 'A-BCD-Extension'

* A-BCD-Extension:
  feat: Add help command to guide user how to use the command

JoanneHing committed:
feat: Create gradle.yml

JoanneHing committed:
Merge commit 'ef242ab6a1d080cde5b8f91f56bf9289cd3a648b'

* commit 'ef242ab6a1d080cde5b8f91f56bf9289cd3a648b':
  feat: Create gradle.yml

JoanneHing committed:
Merge branch 'master' into branch-A-CI

* master:

JoanneHing committed:
Merge branch 'branch-A-CI'

* branch-A-CI:

JoanneHing committed:
feat: Handle invalid timeline error

Timeline error is thrown when
* The timeline entered is earlier than current date and time
* The end time is earlier than begin time for event

JoanneHing committed:
Merge branch 'branch-A-MoreErrorHandling'

* branch-A-MoreErrorHandling:
  feat: Handle invalid timeline error

JoanneHing committed:
test: add JUnit tests for Event, ToDo, and Storage classes

JoanneHing committed:
Merge branch 'branch-A-MoreTesting'

* branch-A-MoreTesting:
  test: add JUnit tests for Event, ToDo, and Storage classes

JoanneHing committed:
test: Add JUnit test to some classes and fix some bug

JoanneHing committed:
Merge branch 'branch-A-MoreTesting'

* branch-A-MoreTesting:
  test: Add JUnit test to some classes and fix some bug

JoanneHing committed:
docs: Add AI tools usage and benefits documentation

JoanneHing committed:
docs: Add user guide documentation

JoanneHing committed:
docs: Update README.md

JoanneHing committed:
docs: Update README.md

JoanneHing committed:
docs: Update README.md

JoanneHing committed:
docs: Update README.md

JoanneHing committed:
style: fix formatting issues

JoanneHing committed:
Merge commit 'dc28c8fa1026a7ab332e271031b9b4da59951cd2'

* commit 'dc28c8fa1026a7ab332e271031b9b4da59951cd2':
  docs: Update README.md
  docs: Update README.md
  docs: Update README.md
  docs: Update README.md

JoanneHing committed:
docs: Update README.md

JoanneHing committed:
Test: Update the JUnit test for Ui

JoanneHing committed:
docs: Fix header comments for the method markUndone

JoanneHing committed:
feat: Add the data file if it does not exist yet

JoanneHing committed:
docs: Update the comment

JoanneHing committed:
docs: Update README.md

JoanneHing committed:
docs: Update comments for certain functions
